# Update python schema model
Only do this if you know what you are doing (so don't).

1. Increase Schema variant number in PEQExcel Input Sheet
2. Save File
3. Open VB Editor
4. Module a_export, Run Macro "Export Schema" by pressing F5
5. In the popup navigate to peexcel/data/exports, press save
6. Check if file is saved in peexcel/data/exports
7. Set correct version = version name in Schema folder code
8. Run Schema folder code
9. Check if folder was created in peexcel/data/schemas/vX_XX_X

# Create Schema folder

In [1]:
# Schema folder code

from pexl.io.converter import xlsx_to_dataset_dir

version = "v1_11_5"

xlsx_to_dataset_dir(
    f"../data/exports/schema_{version}.xlsx", # created by PEExcel macro "export_schema"
    f"../data/schemas/{version}", #creates this directory and csv files
    mode="schema",
    replace_files=True # careful!
)



FileNotFoundError: [Errno 2] No such file or directory: '..\\data\\exports\\schema_v1_11_5.xlsx'

# Schema debugging

In [6]:
from pexl.validate.schema import diff_schema_dirs

diff_result = diff_schema_dirs(
    old_dir=f"../data/schemas/v1.10_4",
    new_dir= f"../data/schemas/{version}"
)
diff_result

{'meta': {'old_dir': '..\\data\\schemas\\v1.10_4',
  'new_dir': '..\\data\\schemas\\v1_11_4',
  'key_col': 'var_name'},
 'tables': {'IN': {'columns': {'added': ['comment',
     'domain',
     'entity_group',
     'entity_key',
     'icon',
     'label_de',
     'measure',
     'spatial_scope',
     'temporal_scope',
     'unit'],
    'removed': ['Einheit', 'Icon', 'Kommentar', 'Name', 'Type'],
    'common_count': 3},
   'rows': {'old_count': 612,
    'new_count': 617,
    'new_errors': 6,
    'var_name_added': ['transmittance_fundament_Wm2NFA',
     'transmittance_heatbridge_Wm2NFA',
     'transmittance_roof_Wm2NFA',
     'transmittance_walls_Wm2NFA',
     'transmittance_windows_Wm2NFA'],
    'var_name_removed': [],
    'duplicate_keys_old': [],
    'duplicate_keys_new': []},
   'changes': {'compared_columns': ['ka', 'Formel'],
    'value_error_count': 6,
    'value_change_count': 67,
    'value_errors': [{'var_name': 'fGHG_grid_column',
      'column': 'Formel',
      'old': 'VERGLEIC

In [7]:
from pexl.validate.schema import schema_diff_to_markdown


md = schema_diff_to_markdown(diff_result, value_format="None")
print(md)

# Schema Diff

- **Old:** `..\data\schemas\v1.10_4`
- **New:** `..\data\schemas\v1_11_4`
- **Key:** `var_name`

## IN

### Columns
- Added: ['comment', 'domain', 'entity_group', 'entity_key', 'icon', 'label_de', 'measure', 'spatial_scope', 'temporal_scope', 'unit']
- Removed: ['Einheit', 'Icon', 'Kommentar', 'Name', 'Type']
- Common count: 3

### Rows
Old count: 612
New count: 617
- **New Error count: 6**
- Added/~~Removed~~ var_name:
   - transmittance_fundament_Wm2NFA
  - transmittance_heatbridge_Wm2NFA
  - transmittance_roof_Wm2NFA
  - transmittance_walls_Wm2NFA
  - transmittance_windows_Wm2NFA



### Potential Errors (#)

| var_name | column | error |
|---|---|---|
| fGHG_grid_column | Formel | VERGLEICH(fGHG_grid_profile;fGHG[#Kopfzeilen];0) |
| fPE_grid_column | Formel | VERGLEICH(fPE_grid_profile;fPE[#Kopfzeilen];0) |
| flex_Signals_selected_column | Formel | VERGLEICH(FLEX_signal_name;Signals[#Kopfzeilen];0) |
| rcp1_dh | Formel | WENNFEHLER(WENN(FINDEN("rneuerbar";SVERWEIS(hea

just copied into a markdown cell:


# Schema Diff

- **Old:** `..\data\schemas\v1.10_4`
- **New:** `..\data\schemas\v1_11_4`
- **Key:** `var_name`

## IN

### Columns
- Added: ['comment', 'domain', 'entity_group', 'entity_key', 'icon', 'label_de', 'measure', 'spatial_scope', 'temporal_scope', 'unit']
- Removed: ['Einheit', 'Icon', 'Kommentar', 'Name', 'Type']
- Common count: 3

### Rows
Old count: 612
New count: 617
- **New Error count: 6**
- Added/~~Removed~~ var_name:
   - transmittance_fundament_Wm2NFA
  - transmittance_heatbridge_Wm2NFA
  - transmittance_roof_Wm2NFA
  - transmittance_walls_Wm2NFA
  - transmittance_windows_Wm2NFA



### Potential Errors (#)

| var_name | column | error |
|---|---|---|
| fGHG_grid_column | Formel | VERGLEICH(fGHG_grid_profile;fGHG[#Kopfzeilen];0) |
| fPE_grid_column | Formel | VERGLEICH(fPE_grid_profile;fPE[#Kopfzeilen];0) |
| flex_Signals_selected_column | Formel | VERGLEICH(FLEX_signal_name;Signals[#Kopfzeilen];0) |
| rcp1_dh | Formel | WENNFEHLER(WENN(FINDEN("rneuerbar";SVERWEIS(heat_th2_carrier_type;cf_constant;7;FALSCH));rcp1_renewable;#NV);rcp1_fossil) |
| rcp2_dh | Formel | WENNFEHLER(WENN(FINDEN("rneuerbar";SVERWEIS(heat_th2_carrier_type;cf_constant;7;FALSCH));rcp2_renewable;#NV);rcp2_fossil) |
| rcp3_dh | Formel | WENNFEHLER(WENN(FINDEN("rneuerbar";SVERWEIS(heat_th2_carrier_type;cf_constant;7;FALSCH));rcp3_renewable;#NV);rcp3_fossil) |





## OUT

### Columns
- Added: ['domain', 'icon', 'ka', 'label_de', 'measure', 'spatial_scope', 'temporal_scope', 'unit']
- Removed: ['Bereich', 'Einheit', 'Formel', 'Kategorie', 'Kommentar', 'Label', 'Name']
- Common count: 1

### Rows
Old count: 410
New count: 410
- **New Error count: 0**
- Added/~~Removed~~ var_name:
 



### Potential Errors (#)
- None

### Value changes
- Compared columns: []
- Error count: 0
- Change count: 0

_No value changes._

## SIM

### Columns
- Added: []
- Removed: []
- Common count: 2

### Rows
Old count: 394
New count: 394
- **New Error count: 3**
- Added/~~Removed~~ var_name:
 



### Potential Errors (#)

| var_name | column | error |
|---|---|---|
| Signal | Formula | =INDEX(Signals,ROW()-ROW(Signals[#Headers]),flex_Signals_selected_column) |
| Ti0cooled | Formula | =IF(NFA_cooled/NFA_total>0,Tsetheat_min,#N/A) |
| Ti0uncooled | Formula | =IF(1-NFA_cooled/NFA_total>0,Tsetheat_min,#N/A) |

### Value changes
- Compared columns: ['Formula']
- Error count: 3
- Change count: 0

_No value changes._

## Audit summary

- Raw audit present (old): True
- Raw audit present (new): True

### Empty var_name exclusions (converter)

**Old:**
```json
{'IN': {'empty_var_name_count': 88}, 'OUT': {'empty_var_name_count': 98}, 'SIM': {'empty_var_name_count': 0}}
```
**New:**
```json
{'IN': {'empty_var_name_count': 82}, 'OUT': {'empty_var_name_count': 98}, 'SIM': {'empty_var_name_count': 0}}
```

In [8]:

from pexl.schema.schema_codegen import generate_schema_module


generate_schema_module(
    excel_path=f"../data/exports/schema_{version}.xlsx",
    output_py_path=f"../src/pexl/schemas/generated/excel_{version.replace(".", "_")}.py",
    version=version.replace(".", "_"), # so filename is valid python module (no dots)
)

import pexl.schema.current as current_schema # latest schema must be manually set
current_schema.ExcelNamedVariables # is used by module excel 

ImportError: cannot import name 'generate_schema_module' from 'pexl.schema.schema_codegen' (C:\Users\schneids\code\peexcel\src\pexl\schema\schema_codegen.py)

In [ ]:
from pexl.io.excel import read_project


project = read_project("../data/exports/ka_project_backup_v1.10.xlsx")
project

In [ ]:
project.districts

In [ ]:
d = project.get("EFHo")
print(d)
d.scenarios

In [ ]:
a = d.scenarios[0]
b = d.get("nach 2020 - existing - Erdgas")
a,b

In [ ]:
print(a.name)
print(a.v.project_description)
print(a.v.GFA_total, a.meta.GFA_total.unit, a.meta.GFA_total)
a.as_dict()